### PR TITLE
Workaround windows lsp args paths

### LINKED ISSUE #6: Why did I not install or start lsp after installation?
image
image
No grammatical hints at all

### PR DESCRIPTION
**Pull request by d1y**
Fix d1y/dockerfile.zed#6 #12013

Co-Authored-By: ZeroHawkeye <166997982@qq.com>

Release Notes:

- Fixed windows lsp-server path

### CODE REVIEW

**Comment by SomeoneToIgnore** [comment]
Could there be more context given? 
It's interesting to know where do such paths come from and why cannot we fix those properly without a workaround.

From my perspective, this PR seems dangerously close to getting closed with a message like https://github.com/zed-industries/zed/pull/11860#issuecomment-2112747921

**Comment by d1y** [comment]
> Could there be more context given?

@ZeroHawkeye It was found that the LSP-server could not start on its Windows machine, and after joint research, it was found that the path path may have been misspelled

https://github.com/zed-industries/zed/blob/73ff7b961617198c187d4ed2ec1a66ebe1a045ce/extensions/vue/src/vue.rs#L75-L83

Among them, env:: current_dir will obtain an absolute path (is cwd, right?)
[join](https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.join) seems to add a prefix before the path`/`

```
/C:\Users\zero\AppData\Local\Zed\extensions\work\dockerfile/node_modules/dockerfile-language-server-nodejs/bin/docker-langserver
```

After being handed over to node to run, this path will become(https://github.com/d1y/dockerfile.zed/issues/6#issuecomment-2119146389)

```
E:\C:\Users\zero\AppData\Local\Zed\extensions\work\dockerfile/node_modules/dockerfile-language-server-nodejs/bin/docker-langserver
```

The solution I can think of is to remove the prefix `/` (https://github.com/d1y/dockerfile.zed/issues/6#issuecomment-2119158355)

After all, we cannot ask the extension author to fix this path issue themselves? 🤔

------------

If my understanding is correct, perhaps changing the join in the extension to [path.canonicalize](https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.canonicalize)? <[examples processed?](https://github.com/zed-industries/zed/pull/11083)>

**Comment by JunkuiZhang** [comment]
Hey, this may not be a Windows related issue, I can install both dockerfile and vue extentions. 

Am I downloading the right extensions?

Screenshot 2024-05-20 160350


Screenshot 2024-05-20 160405

**Comment by ZeroHawkeye** [comment]
> 嘿，这可能不是 Windows 相关的问题，我可以安装 dockerfile 和 vue 扩展。
> 
> 不熟悉 vue，我是否下载了正确的扩展？
> 
> 截图2024-05-20 160350
> 
> 截图2024-05-20 160405

目前确实可以下载扩展，不清楚您这边可以正常运行docker，vue等插件的lsp提示么？我这边windows10会因为路径错误启动lsp失败

**Comment by JunkuiZhang** [comment]
> 目前确实可以下载扩展，不清楚您这边可以正常运行docker，vue等插件的lsp提示么？我这边windows10会因为路径错误启动lsp失败

Ahh, no.
```rust
 Cannot find module 'C:\\C:\\Users\..
```

File: crates/extension/src/extension_lsp_adapter.rs [hunk 1]
```diff
@@ -94,7 +94,24 @@ impl LspAdapter for ExtensionLspAdapter {
...
+                    .map(|arg| {
+                        #[cfg(windows)]
+                        {
+                            use regex::Regex;
+                            // Workaround windows path
+                            // `/C:\\Users\\d1y` need remove `/` prefix
+                            let pattern = r"^/[A-Z]:\\";
+                            let regex = Regex::new(pattern).unwrap();
+                            if regex.is_match(&arg) {
+                                return arg.trim_start_matches('/').into();
```
Comment (huacnlee) [inline crates/extension/src/extension_lsp_adapter.rs:109]: I think, the Vue not work, may not cause by this.

image

It's looks like this line will let the Vue server bin path to path of the `Zed` started path.

image

As you see on my log output, the `@vue\\language-server\\bin\\vue-language-server.js` has  `F:\\` prefix. That is the Zed.exe I was installed to: `F:\Zed\Zed.exe`

**Comment by huacnlee** [comment]
I think, the Vue not work, may not cause by this.

image

It's looks like this line will let the Vue server bin path to path of the `Zed` started path.

image

As you see on my log output, the `@vue\\language-server\\bin\\vue-language-server.js` has  `F:\\` prefix. That is the Zed.exe I was installed to: `F:\Zed\Zed.exe`

**Comment by ZeroHawkeye** [comment]
> 动路径的路

zed is lsp installed under $user-AppData. I tested zed by running it under c drive and opening the e drive project. The result is that the lsp prefix is ​​E:--C:--...

**Comment by tommy141x** [comment]
This is my log attempting to compile & run this PR:

`[2024-05-22T20:06:38-04:00 ERROR util] crates\lsp\src/lsp.rs:357: cannot read LSP message headers
[2024-05-22T20:06:38-04:00 ERROR project] failed to start language server "tailwindcss-language-server": oneshot canceled
[2024-05-22T20:06:38-04:00 ERROR project] server stderr: Some("C:\\Users\\tommy\\AppData\\Local\\Zed\\languages\\tailwindcss-language-server\\node_modules\\.bin\\tailwindcss-language-server:2\r\nbasedir=$(dirname \"$(echo \"$0\" | sed -e 's,\\\\,/,g')\")\r\n          ^^^^^^^\r\n\r\nSyntaxError: missing ) after argument list\r\n    at internalCompileFunction (node:internal/vm:73:18)\r\n    at wrapSafe (node:internal/modules/cjs/loader:1176:20)\r\n    at Module._compile (node:internal/modules/cjs/loader:1218:27)\r\n    at Module._extensions..js (node:internal/modules/cjs/loader:1308:10)\r\n    at Module.load (node:internal/modules/cjs/loader:1117:32)\r\n    at Module._load (node:internal/modules/cjs/loader:958:12)\r\n    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)\r\n    at node:internal/main/run_main_module:23:47\r\n\r\nNode.js v18.15.0\r\n")
[2024-05-22T20:06:38-04:00 ERROR util] crates\lsp\src/lsp.rs:773: server shut down
[2024-05-22T20:06:39-04:00 ERROR util] crates\languages\src\tailwind.rs:229: missing executable in directory "C:\\Users\\tommy\\AppData\\Local\\Zed\\languages\\tailwindcss-language-server\\node_modules"
[2024-05-22T20:06:39-04:00 ERROR util] crates\lsp\src/lsp.rs:357: cannot read LSP message headers
[2024-05-22T20:06:39-04:00 ERROR project] failed to start language server "tailwindcss-language-server": oneshot canceled
[2024-05-22T20:06:39-04:00 ERROR project] server stderr: Some("node:internal/modules/cjs/loader:1078\r\n  throw err;\r\n  ^\r\n\r\nError: Cannot find module 'C:\\Users\\tommy\\AppData\\Local\\Zed\\languages\\tailwindcss-language-server\\node_modules\\.bin\\tailwindcss-language-server'\r\n    at Module._resolveFilename (node:internal/modules/cjs/loader:1075:15)\r\n    at Module._load (node:internal/modules/cjs/loader:920:27)\r\n    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)\r\n    at node:internal/main/run_main_module:23:47 {\r\n  code: 'MODULE_NOT_FOUND',\r\n  requireStack: []\r\n}\r\n\r\nNode.js v18.15.0\r\n")
[2024-05-22T20:06:39-04:00 ERROR util] crates\lsp\src/lsp.rs:773: server shut down
[2024-05-22T20:06:40-04:00 ERROR util] crates\languages\src\tailwind.rs:229: The system cannot find the path specified. (os error 3)
[2024-05-22T20:06:40-04:00 ERROR util] crates\language\src\language_registry.rs:877: server container removal

Caused by:
    The system cannot find the file specified. (os error 2)`

**Comment by Ianec21** [comment]
thanks bro, works like a charm

**Comment by d1y** [comment]
As no one has been handling this, I have decided to close this PR

**Pull request closed without merging**